File: test/stable_hash_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:cookrange/core/utils/stable_hash.dart';

// Faz A §A3 — a shared vector table, asserted here AND in
// functions/test/stable_hash.test.js against the SAME expected outputs.
// That duplication is the point: it is what proves the Dart and Node
// implementations agree, which is the entire reason this hash replaced
// String.hashCode (a Cloud Function must reproduce the same rollout
// bucket a client computed). The first three vectors are the well-known
// public FNV-1a-32 reference vectors (independent confirmation this is a
// textbook-correct implementation, not just internally self-consistent);
// the rest are realistic `"$feature:$uid"`-shaped inputs mirroring
// AppConfigService.isInRollout's actual usage, plus edge cases (empty
// string, non-ASCII/Turkish text, long input, short numeric-looking
// strings that could collide if truncation were wrong).
void main() {
  final vectors = <String, int>{
    '': 2166136261,
    'a': 3826002220,
    'foobar': 3214735720,
    'gym:abc123': 2999454414,
    'coach:xyz789': 1718838574,
    'squad:uid_0001': 3313215373,
    'programs:': 1538371286,
    ':uid': 1822523445,
    'gym_invite_codes:9f8e7d6c5b4a3210': 3511076557,
    'a' * 100: 168538585,
    'Türkçe karakterler: çğıöşü': 3039875672,
    '0': 890022063,
    '00': 569209421,
    'meal_plan_templates:uidWithMixedCASE123': 820354505,
    'gym_attribution:uid-with-dashes-999': 152514592,
  };

  group('fnv1a32 — shared vector table (cross-checked against functions/test/stable_hash.test.js)', () {
    vectors.forEach((input, expected) {
      final label = input.length > 24 ? '${input.substring(0, 24)}…' : input;
      test('"$label" (len ${input.length}) -> $expected', () {
        expect(fnv1a32(input), expected);
      });
    });
  });

  group('fnv1a32 — properties', () {
    test('always returns a value in the unsigned 32-bit range', () {
      for (final input in vectors.keys) {
        final h = fnv1a32(input);
        expect(h, greaterThanOrEqualTo(0));
        expect(h, lessThanOrEqualTo(0xFFFFFFFF));
      }
    });

    test('is deterministic across repeated calls', () {
      const input = 'gym:repeat-me';
      final first = fnv1a32(input);
      for (var i = 0; i < 5; i++) {
        expect(fnv1a32(input), first);
      }
    });

    test('different inputs produce different hashes (no trivial collision '
        'among these vectors)', () {
      final hashes = vectors.values.toSet();
      expect(hashes.length, vectors.length);
    });
  });
}

File: functions/test/stable_hash.test.js
'use strict';

// Faz A §A3 — shared vector table, asserted here AND in
// test/stable_hash_test.dart against the SAME expected outputs. See that
// file's header comment for the full rationale; the short version: a
// Cloud Function must reproduce the exact rollout bucket a client
// computed for AppConfigService.isInRollout, which String.hashCode could
// not guarantee. This pair of tests is what proves the two independent
// implementations (Dart, Node) actually agree.
//
//   node --test functions/test/stable_hash.test.js

const { test, describe } = require('node:test');
const assert = require('node:assert/strict');
const { fnv1a32 } = require('../stable_hash');

const vectors = {
  '': 2166136261,
  a: 3826002220,
  foobar: 3214735720,
  'gym:abc123': 2999454414,
  'coach:xyz789': 1718838574,
  'squad:uid_0001': 3313215373,
  'programs:': 1538371286,
  ':uid': 1822523445,
  'gym_invite_codes:9f8e7d6c5b4a3210': 3511076557,
  [('a').repeat(100)]: 168538585,
  'Türkçe karakterler: çğıöşü': 3039875672,
  0: 890022063,
  '00': 569209421,
  'meal_plan_templates:uidWithMixedCASE123': 820354505,
  'gym_attribution:uid-with-dashes-999': 152514592,
};

describe('fnv1a32 — shared vector table (cross-checked against test/stable_hash_test.dart)', () => {
  for (const [input, expected] of Object.entries(vectors)) {
    const label = input.length > 24 ? `${input.slice(0, 24)}…` : input;
    test(`"${label}" (len ${input.length}) -> ${expected}`, () => {
      assert.equal(fnv1a32(input), expected);
    });
  }
});

describe('fnv1a32 — properties', () => {
  test('always returns a value in the unsigned 32-bit range', () => {
    for (const input of Object.keys(vectors)) {
      const h = fnv1a32(input);
      assert.ok(h >= 0);
      assert.ok(h <= 0xFFFFFFFF);
    }
  });

  test('is deterministic across repeated calls', () => {
    const input = 'gym:repeat-me';
    const first = fnv1a32(input);
    for (let i = 0; i < 5; i++) {
      assert.equal(fnv1a32(input), first);
    }
  });

  test('different inputs produce different hashes (no trivial collision among these vectors)', () => {
    const hashes = new Set(Object.values(vectors));
    assert.equal(hashes.size, Object.keys(vectors).length);
  });
});
